### DISCUSSION TITLE
CUDA  question

### DISCUSSION THREAD

**Discussion by ljeft** [idaholab/moose#26233] [Q&A General] [answered]
Hello Mooser
I know that Moose can improve efficiency by calling GPU through Pestc, now I wonder where Moose wrote CUDA-related code, I am trying to port CUDA code. Allow Moose to call DCU (Datacenter Acceleration Unit)。

**Comment by lindsayad** [comment]
I'm not quite sure what you're asking. MOOSE does not have much CUDA-related code inside of the repository. The only bit is in `BatchMaterial.h` written by @dschwen

**Reply by ljeft** [reply]
Sorry, maybe I don't know enough about moose, I just want to try running moose using gpu, I just watched moose's YouTube video "Porting PETSc-based application codes to GPU-accelerated architectures" .This question was raised before I discovered this video.

**Reply by GiudGiud** [reply]
Hello

I think porting all of MOOSE might be too much work.
As Alex mentioned materials can be ported to GPU.
Another project would be to make a generic GPU PetscApp and add the links from this app to MOOSE in the external_petsc_solver MOOSE module.

Guillaume

**Reply by ljeft** [reply]
Can you give me a case  or a link to this project?

**Reply by GiudGiud** [reply, marked as answer]
the main class is there
https://mooseframework.inl.gov/source/problems/ExternalPETScProblem.html

I dont know if there's much documentation outside the code.
There is no ongoing work there, it s just an idea for integrating GPU code with moose

**Reply by lindsayad** [reply]
You may also want to read https://petsc.org/release/faq/#doc-faq-gpuhowto. Something we would want to do to help MOOSE applications call through to PETSc with GPU support is to leverage the PETSc GPU Vec and Mat types. For instance we could programatically create PETSc CUDA vectors and matrices. This would start at the libMesh level. Just grepping in libMesh for `VECSEQCUDA` or `VECMPICUDA` yields no hits.

~I think we have not spent a lot of time on this yet because it's been unclear what GPU implementations would become most used. E.g. with limited resources should we focus on CUDA, OpenCL, something else? Anyway if you just want to get through to PETSc, then the external problem that @GiudGiud pointed to is a good route.~

FYI @roystgnr

**Reply by lindsayad** [reply]
There is actually a (newly again) active Discussion about GPUs over at https://github.com/idaholab/moose/discussions/21152. I've struck through my second paragraph above. If you have a GPU enabled PETSc, you should be able to perform GPU solves by passing GPU types for the PETSc options `-vec_type` and `-mat_type`. And you should be able to try this with any MOOSE input

**Reply by roystgnr** [reply]
I don't think "E.g. with limited resources should we focus on CUDA, OpenCL, something else?" should have been struck - IMHO it's *the* huge sticking point.  Trying to mix CPU assembly with GPU solve gets hobbled by RAM->VRAM bandwidth on most architectures, so we really want GPU assembly too.  (although, it's been forever since I tested that; I should see how the `$LIBMESH_BENCHMARK` run goes if I just set `-vec_type` and `-mat_type`...)

My original libMesh-on-GPUs experiment used VexCL, which was fantastic for everything it did but which doesn't let you go to *completely* general kernels without ending up writing raw GPU code anyway.  A colleague in a GPGPU-intensive job tells me that the most popular alternatives today are (oversimplifying and paraphrasing): Kokkos (which nobody likes), CUDA (which is slightly less ugly but cuts your hardware compatibility in half), or CUDA+HIP (which is slightly less ugly but twice as much work).  I did some playing with the Nvidia HPC SDK a year or so ago, and that amazingly lets you write a lot of GPGPU code via "C++17 plus a magic compiler flag" on recent Nvidia cards, but IIRC when I dropped that I'd already reported two showstopper bugs and one of them hadn't yet been fixed.
